crates/network/README.md

```diff
@@ -0,0 +1,302 @@
+
+#### mDNS
+
+- Purpose: Local network peer discovery
+- Key Events:
+  - `Discovered`: New peers found on the local network
+
+#### Rendezvous
+
+- Purpose: Peer discovery through a known rendezvous point
+- Key Events:
+  - `Discovered`: Peers found through the rendezvous server
+  - `RegisterFailed`/`DiscoverFailed`: Issues with rendezvous operations
+
+### Data Exchange Protocols
+
+#### Gossipsub
+
+- Purpose: Efficient multi-peer data broadcasting
+- Every `context` is one gossip topic
+- Key Events:
+  - `Message`: New message received on a subscribed topic
+  - `Subscribed`: Successfully subscribed to a topic
+
+### Connectivity Protocols
+
+#### Relay
+
+- Purpose: Enables communication between peers that can't directly connect
+
+#### DCUtR (Direct Connection Upgrade through Relay)
+
+- Purpose: Helps establish direct connections between peers initially connected via a relay
+
+### Meta Protocols
+
+#### Identify
+
+- Purpose: Exchange of node metadata and supported protocols
+- Key Events:
+  - `Received`: Received identify information from a peer
+
+#### Ping
+
+- Purpose: Keep-alive and latency measurements
+
+### Custom Protocol
+
+#### Stream
+
+- Purpose: Application-specific data exchange
+- Implementation: Uses a custom `MessageJsonCodec` for serialization/deserialization
+- Features:
+  - Bidirectional communication between peers
+  - Message Structure: The protocol uses a custom `Message` struct defined as
+  
+ ```rust
+  pub  struct  Message  { pub data:  Vec<u8> }
+```
+
+This flexible structure allows for sending arbitrary data between peers.
+
+- Key Events:
+  - Handled directly in the `EventLoop` as `StreamOpened` events
+
+The custom stream protocol allows for tailored communication patterns specific to our application needs, complementing the standardized libp2p protocols.
+
+## Discovery
+
+We combine multiple protocols in order to improve our connectivity.
+
+- At the begging we dial Calimero boot nodes (which speak Calimero KAD protocol) and we start discovering peers on local network with mDNS.
+
+- During identify exchange (which occurs for every established connection) we record protocols which other peer supports.
+
+- For discovered peers with mDNS we perform direct dial (because we want to be connected to all local peers).
+
+- For discovered peers with rendezvous we perform dial only if peer is not already connected (this means that it most likely already discovered and dialed due to mDNS event).
+
+Beside network state by lip2p (more in `self.swarm.network_info()`) we also have our discovery state.
+
+We keep track of multiaddrs for all connected peers and peers of interest (which are never removed from state so we can reconnect to them when needed).
+
+To improve connectivity, as we discover relay nodes we attempt to make a relay reservation (so other peers can hole punch us).
```

Comment:
Hole punching is one of the considerations for using a relay, but just being exposed via a relay in itself is a net benefit for connectivity